go/plugins/firebase/retriever.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firebase
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/firestore"
+	"github.com/firebase/genkit/go/ai"
+)
+
+const provider = "firebase"
+
+// RetrieverOptions defines the configuration for the retriever.
+type RetrieverOptions struct {
+	Name            string
+	Label           string
+	Client          *firestore.Client
+	Embedder        ai.Embedder
+	EmbedderOptions ai.EmbedOption
+	Collection      string
+	VectorField     string
+	MetadataFields  []string // Optional: if empty, metadata will not be retrieved
+	ContentField    string
+	DistanceMeasure firestore.DistanceMeasure
+}
+
+type RetrieverRequestOptions struct {
+	Limit           int `json:"limit,omitempty"` // maximum number of values to retrieve
+	DistanceMeasure firestore.DistanceMeasure
+}
+
+func DefineFirestoreRetriever(cfg RetrieverOptions) (ai.Retriever, error) {
+
+	Retrieve := func(ctx context.Context, req *ai.RetrieverRequest) (*ai.RetrieverResponse, error) {
+		if req == nil {
+			return nil, fmt.Errorf("retriever request is nil")
+		}
+
+		options := RetrieverRequestOptions{Limit: 10, DistanceMeasure: cfg.DistanceMeasure}
+
+		if req.Options != nil {
+			// Ensure that the options are of the correct type
+			parsedOptions, ok := req.Options.(RetrieverRequestOptions)
+			if !ok {
+				return nil, fmt.Errorf("firebase.Retrieve options have type %T, want %T", req.Options, &RetrieverRequestOptions{})
```

Comment:
Do you want a struct (line 57) or a pointer to a struct (this line)?

Reply:
(My college roommate pulled an all-nighter writing C code. In the morning he said, "I'm so tired, I don't know whether I want a shower or a pointer to a shower.")